src/restart_windows.nim

```diff
@@ -0,0 +1,86 @@
+        szExeFile: array[MAX_PATH, cchar]
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot
+proc createToolhelp32Snapshot(dwFlags: DWORD, th32ProcessID: DWORD): Handle
+    {.stdcall, dynlib: "kernel32", importc: "CreateToolhelp32Snapshot".}
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-process32first
+proc process32First(hSnapshot: Handle, lppe: var PROCESSENTRY32): bool
+    {.stdcall, dynlib: "kernel32", importc: "Process32First".}
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-process32next
+proc process32Next(hSnapshot: Handle, lppe: var PROCESSENTRY32): bool
+    {.stdcall, dynlib: "kernel32", importc: "Process32Next".}
+
+## Unlike POSIX getppid, can be passed an arbitrary PID.
+proc getppid(pid = getCurrentProcessId().DWORD): DWORD =
+    # dwFlags=2 causes the snapshot to include all currently running processes
+    # th32ProcessID=0 refers to the calling process
+    let handle = createToolhelp32Snapshot(dwFlags = 2, th32ProcessID = 0)
+    var processEntry: PROCESSENTRY32
+    processEntry.dwSize = sizeof(PROCESSENTRY32).DWORD
+
+    if process32First(handle, processEntry):
+        doWhile process32Next(handle, processEntry):
+            if processEntry.th32ProcessID == pid:
+                discard closeHandle handle
+                return processEntry.th32ParentProcessID
+
+## Invoke the native messenger again with special arguments and outside the
+## Job Firefox starts it in. This is necessary because Firefox kills every
+## process in the messenger's job when it itself exits, but to be able to
+## restart Firefox we need a process which survives that.
+proc cloneMessenger*(profiledir, browsername: string) =
+    var startupInfo: STARTUPINFO
+    var processInformation: PROCESS_INFORMATION
+    discard createProcessW(
+        nil.newWideCString,
+        quoteShellCommand([getAppFilename(), "restart", $ getppid(), profiledir,
+            browsername]).newWideCString,
+        nil,
+        nil,
+        false.WINBOOL,
+        # dwFlags = CREATE_BREAKAWAY_FROM_JOB | CREATE_NO_WINDOW
+        # See https://docs.microsoft.com/en-gb/windows/win32/procthread/process-creation-flags#CREATE_BREAKAWAY_FROM_JOB
+        0x01000000 or 0x08000000,
+        nil,
+        nil,
+        startupInfo,
+        processInformation
+    )
+
+## The "main function" for the cloned messenger, responsible for the actual
+## restarting. Please rename this if you can think of a better name.
+proc restartWindowsStep2*(firefoxPid: int, profiledir, browsername: string) =
+    let firefoxHandle = openProcess(
+        dwDesiredAccess = SYNCHRONIZE.DWORD,
+        bInheritHandle = false.WINBOOL,
+        dwProcessId = firefoxPid.DWORD
+    )
+    discard waitForSingleObject(
+        hHandle = firefoxHandle,
+        dwMilliseconds = INFINITE
+    )
```

Comment:
```suggestion
proc launchFirefoxAtPid*(firefoxPid: int, profiledir, browsername: string) =
    return openProcess(
        dwDesiredAccess = SYNCHRONIZE.DWORD,
        bInheritHandle = false.WINBOOL,
        dwProcessId = firefoxPid.DWORD,
    )


# Somewhere else...
discard waitForSingleObject(
    hHandle = launchFirefoxAtPid(...),
    dwMilliseconds = INFINITE,
)
```


`profiledir` and `browsername` aren't used - is that on purpose?

Reply:
This part of the code isn't finished yet – this doesn't actually restart Firefox yet, just waits for it to exit. Once it does, it'll need those params.